### ISSUE TITLE
Results XML is occasionally incomplete

### ISSUE THREAD

**Issue by dougbu** [xunit/xunit#1534]
1. Clone the https://github.com/aspnet/AspNetWebStack repo
2. `cd AspNetWebStack`
3. `.\build.cmd`
4. Look for warnings at the end of the build, from the `PrintTestRunSummary` target
5. If none are output, goto 3

#### Expected
The above loop should be pretty much endless, never seeing `PrintTestRunSummary` warnings. (See this [relatively-new block](https://github.com/aspnet/AspNetWebStack/blob/7674394480666e66830689d07002e3a2e26ebc32/tools/WebStack.tasks.targets#L181-L192) to output those warnings.)

#### Actual
Usually only need to loop a few times before warnings appear. A varying results file is affected when warnings do occur. Size of the results file and duration of the test run doesn't seem to matter. Every so often, two or three results files are incomplete.

The test run for the incomplete results does not indicate any issues i.e. the problem is visible only later, when `PrintTestRunSummary` examines the results file. Test execution summary is also normal e.g. 
```
    Finished:    Microsoft.Web.Helpers.Test
  === TEST EXECUTION SUMMARY ===
     Microsoft.Web.Helpers.Test  Total: 162, Errors: 0, Failed: 0, Skipped: 0, Time: 11.500s
```

### Notes
This issue is inconsistent but may occur more when the system is loaded. For example, testing in parallel (the default) makes the problem occur more often. Serial testing (`/p:TestInParallel=false` on the `.\build.cmd` command line) makes the build almost-but-not-quite completely reliable on an otherwise lightly-used system.

Have seen the warnings in `CodeAnalysis` (the default), `Debug` and `Release` builds. Might occur more often with `Debug` i.e. `.\build.cmd UnitTest /p:Configuration=Debug`.

This repo uses the MSBuild runner in command-line builds. But, wouldn't be too surprising if the Console runner was also affected.

This problem is _not_ related to #1192. The results file is valid, not truncated.

Appears like `DelegatingXmlCreationSink.HandleTestAssemblyStarting(...)` is called but `HandleTestAssemblyFinished(...)` is not.

#### Partial results
``` xml
<?xml version="1.0" encoding="utf-8"?>
<assemblies timestamp="10/26/2017 09:04:34">
  <assembly name="C:\dd\dnx\Universe++\AspNetWebStack\bin\Debug\test\Microsoft.Web.Helpers.Test.DLL" environment="32-bit .NET 4.0.30319.42000 [collection-per-class, parallel (4 threads)]" test-framework="xUnit.net 2.3.0.3820" run-date="2017-10-26" run-time="09:04:30" config-file="C:\Program Files (x86)\Microsoft Visual Studio\Preview\INT\MSBuild\15.0\Bin\MSBuild.exe.Config">
    <errors />
  </assembly>
</assemblies>
```

#### Full results
[Microsoft.Web.Helpers.Test-XunitResults (full).txt](https://github.com/xunit/xunit/files/1419309/Microsoft.Web.Helpers.Test-XunitResults.full.txt) (file actually XML)

**Comment by BrunoJuchli** [comment]
We are experiencing the same issue. xunit.console.exe File version 2.3.1.3858.
We call xunit with the following arguments:

     xunit.console.exe XYZ.dll -nologo -noshadow -noappdomain -xml XYZ.test -parallel none

Notice, that in the produced xml, the `<assembly>` element is missing all attributes after `config-file`. If everything works as it should, there would be the number of tests/failures etc. after the `config-file` attribute.

**Comment by wireddown** [comment]
My engineering team is also impacted by this bug. After xUnit exits successfully, the XML file that it writes, while valid with matching tags, is missing every result.

# Details
```
$ "C:\Users\Public\Documents\Scope driver tests\xunit.console.exe" IFNIScopeHardwareTest.dll -nunit "C:\Users\Public\Documents\Scope driver tests\TestResults.xml"
xUnit.net Console Runner (64-bit Desktop .NET 4.0.30319.42000)
 Discovering: IFNIScopeHardwareTest
 Discovered: IFNIScopeHardwareTest
 Starting: IFNIScopeHardwareTest
 Finished: IFNIScopeHardwareTest
=== TEST EXECUTION SUMMARY ===
 IFNIScopeHardwareTest Total: 135, Errors: 0, Failed: 0, Skipped: 0, Time: 236.761s

$ cat "C:\Users\Public\Documents\Scope driver tests\TestResults.xml"
\xef\xbb\xbf<?xml version="1.0" encoding="utf-8"?>
<test-results name="Test results" errors="0" inconclusive="0" ignored="0" invalid="0" not-run="0" date="2019-12-31" time="19:21:16" total="0" failures="0" skipped="0">
 <environment os-version="unknown" platform="unknown" cwd="unknown" machine-name="unknown" user="unknown" user-domain="unknown" nunit-version="xUnit.net 2.3.1.3858" clr-version="64-bit .NET 4.0.30319.42000 [collection-per-class, parallel (8 threads)]" />
 <culture-info current-culture="unknown" current-uiculture="unknown" />
 <test-suite type="Assemblies" name="xUnit.net Tests" executed="True" success="True" result="Success" time="0">
 <results>
 <test-suite type="Assembly" executed="True" name="C:\\Users\\Public\\Documents\\Scope driver tests\\IFNIScopeHardwareTest.DLL" result="" success="" time="">
 <results />
 </test-suite>
 </results>
 </test-suite>
</test-results>
```

**Comment by bradwilson** [comment]
I'm assuming this was caused by the race condition that was fixed by `ExecutionSink` in [2.6.3](https://xunit.net/releases/v2/2.6.3). If you're still experiencing this issue after upgrading, please open a new issue with a repro project. Thanks!
